src/Yarhl/IO/DataStream.cs

```diff
@@ -581,7 +597,7 @@ private void IncreaseStreamCounter()
             }
 
             if (!Instances.ContainsKey(BaseStream)) {
-                Instances.Add(BaseStream, 1);
+                Instances.TryAdd(BaseStream, 1);
```

Comment:
This `TryAdd` may fail if two thread evaluate at the same time that the dictionary doesn't contain the `BaseStream` and then only one of them will success in the add. In that case the other one instead of failing and giving up should increase the counter by one. So... I think we could fix this scenario by checking the result of `TryAdd` and if it fails it means it was added and we need to increase the counter.

Reply:
I think we would need to take the lock of the `BaseStream` here too for the whole if, and also in the `Dispose` when we check to remove because otherwise a `DataStream` may delete it while another one may try to increase it.